pkg/operator/cluster/ceph/osd/pod.go

```diff
@@ -40,70 +44,154 @@ const (
+		dataDirSource := v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: c.dataDirHostPath}}
+		volumes = append(volumes, v1.Volume{Name: k8sutil.DataDirVolume, VolumeSource: dataDirSource})
+	} else {
+		logger.Infof("no data dir provided")
+	}
+
+	// by default, don't define any volume config unless it is required
+	if len(devices) > 0 || selection.MetadataDevice != "" {
+		// create volume config for the data dir and /dev so the pod can access devices on the host
+		devVolume := v1.Volume{Name: "devices", VolumeSource: v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: "/dev"}}}
+		volumes = append(volumes, devVolume)
+		devMount := v1.VolumeMount{Name: "devices", MountPath: "/dev"}
+		volumeMounts = append(volumeMounts, devMount)
+	}
 
-	return &extensions.ReplicaSet{
+	if len(volumes) == 0 {
+		logger.Warningf("empty volumes")
+		return nil
+	}
+
+	envVars := []v1.EnvVar{
+		nodeNameEnvVar(),
+		k8sutil.PodIPEnvVar(k8sutil.PrivateIPEnvVar),
+		k8sutil.PodIPEnvVar(k8sutil.PublicIPEnvVar),
+	}
+
+	privileged := true
+	runAsUser := int64(0)
+	readOnlyRootFilesystem := false
+	DNSPolicy := v1.DNSClusterFirst
+	if c.HostNetwork {
+		DNSPolicy = v1.DNSClusterFirstWithHostNet
+	}
+	journalStr := ""
+	if osd.IsFileStore {
+		journalStr = fmt.Sprintf("--osd-journal=%s", osd.Journal)
+	}
+	return &extensions.Deployment{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:            fmt.Sprintf(appNameFmt, nodeName),
+			Name:            fmt.Sprintf(osdAppNameFmt, osd.ID),
 			Namespace:       c.Namespace,
 			OwnerReferences: []metav1.OwnerReference{c.ownerRef},
 			Labels: map[string]string{
 				k8sutil.AppAttr:     appName,
 				k8sutil.ClusterAttr: c.Namespace,
+				osdLabelKey:         fmt.Sprintf("%d", osd.ID),
 			},
 		},
-		Spec: extensions.ReplicaSetSpec{
-			Template: podSpec,
+		Spec: extensions.DeploymentSpec{
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: appName,
+					Labels: map[string]string{
+						k8sutil.AppAttr:     appName,
+						k8sutil.ClusterAttr: c.Namespace,
+						osdLabelKey:         fmt.Sprintf("%d", osd.ID),
+					},
+					Annotations: map[string]string{},
+				},
+				Spec: v1.PodSpec{
+					NodeSelector: map[string]string{apis.LabelHostname: nodeName},
+					//FIXME: use a diff SA for osd daemon
+					// ServiceAccountName: appName,
+					RestartPolicy: v1.RestartPolicyAlways,
+					HostNetwork:   c.HostNetwork,
+					DNSPolicy:     DNSPolicy,
+					Containers: []v1.Container{
+						{
+							Command: []string{"sh", "-c",
+								fmt.Sprintf("ceph-osd --foreground --id %d --conf %s --osd-data %s --keyring %s --cluster %s --osd-uuid %s %s %s %s",
+									osd.ID,
+									osd.Config,
+									osd.DataPath,
+									osd.KeyringPath,
+									osd.Cluster,
+									osd.UUID,
+									fmt.Sprintf("--public-addr=${%s}", k8sutil.PublicIPEnvVar),
+									fmt.Sprintf("--cluster-addr=${%s}", k8sutil.PrivateIPEnvVar),
```

Comment:
What about using Kubernetes environment variable substitution here so `sh` is not needed? https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/#use-environment-variables-to-define-arguments

Reply:
yes, the env var is already set but `ceph-osd` command itself doesn't check env var. So I have to use both env var and `sh` to resolve the env var and flatten into a well resolved command for `ceph-osd`

Reply:
But it isn't `ceph-osd` that has to resolve/check the env var, Kubernetes does resolve it to the value already.
With env var and `sh` it ends up with for the `ceph-osd` command flag `--cluster-addr=10.10.10.10`, with Kubernetes env var substitution it would result in the same, wouldn't it?

Reply:
here is the spec
```
# kubectl describe pod  -n rook rook-ceph-osd-id-0-76748cddf6-4r9b6
Name:           rook-ceph-osd-id-0-76748cddf6-4r9b6
Namespace:      rook
Node:           127.0.0.1/127.0.0.1
Start Time:     Fri, 25 May 2018 11:59:17 -0400
Labels:         app=rook-ceph-osd
                pod-template-hash=3230478892
                rook.io/ceph-osd-id=0
                rook_cluster=rook
Annotations:    <none>
Status:         Running
IP:             172.17.0.8
Controlled By:  ReplicaSet/rook-ceph-osd-id-0-76748cddf6
Containers:
  rook-ceph-osd:
    Container ID:  docker://ec805a9fab1fe63940920b4a8f09d00597248158863b94bcd29ee310f7aa0cec
    Image:         rook/rook:master
    Image ID:      docker-pullable://localhost:5000/rook@sha256:ba2207d8c0317f40aa268e3ec182fac772d9a4942b6a579a67b926a95cd0f0f1
    Port:          <none>
    Host Port:     <none>
    Command:
      ceph-osd
    Args:
      --foreground
      --id
      0
      --conf
      /var/lib/rook/filestore/osd0/rook.config
      --osd-data
      /var/lib/rook/filestore/osd0
      --keyring
      /var/lib/rook/filestore/osd0/keyring
      --cluster
      rook
      --osd-uuid
      119e5be9-5941-491a-b52a-cf9f6b5dc088
      --public-addr=${ROOK_PUBLIC_IPV4}
      --cluster-addr=${ROOK_PRIVATE_IPV4}
      --osd-journal=/var/lib/rook/filestore/osd0/journal
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Fri, 25 May 2018 12:00:49 -0400
      Finished:     Fri, 25 May 2018 12:00:50 -0400
    Ready:          False
    Restart Count:  4
    Limits:
      cpu:     0
      memory:  0
    Requests:
      cpu:     0
      memory:  0
    Environment:
      ROOK_NODE_NAME:      (v1:spec.nodeName)
      ROOK_PRIVATE_IPV4:   (v1:status.podIP)
      ROOK_PUBLIC_IPV4:    (v1:status.podIP)
``

Reply:
But you used `{` and `}` brackets which is "POSIX syntax", Kubernetes env var substitution uses `(` and `)` brackets.
In my example using `$(TEST_ENV_VAR)` in a `command` and/or `args` works.
This is my test Pod manifest:
```
apiVersion: v1
kind: Pod
metadata:
  name: busybox
spec:
  containers:
  - image: busybox
    env:
    - name: MESSAGE
      value: "hello world"
    command: ["/bin/echo"]
    args: ["$(MESSAGE)"]
    name: busybox
  restartPolicy: Never
```
After it ran:
```
$ kubectl logs busybox
hello world
```
So Kubernetes substituted the `$(MESSAGE)` with the env var correctly.

Reply:
@galexrt no, not working. I tried both ways with `"--public_addr=$(%s)"` and just `"$(%s)"`

It is probably due to the fact the env var is a `value_from` (i.e. from `v1:status.podIP`)

```
Containers:
    Command:
      ceph-osd
    Args:
      --foreground
      --id
      0
      --conf
      /var/lib/rook/osd0/rook.config
      --osd-data
      /var/lib/rook/osd0
      --keyring
      /var/lib/rook/osd0/keyring
      --cluster
      rook
      --osd-uuid
      a975f51f-a7d3-4282-939d-46d83fdb6b05
      --public-addr=$(ROOK_PUBLIC_IPV4)
      --cluster-addr=$(ROOK_PRIVATE_IPV4)
      
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Tue, 29 May 2018 09:32:58 -0400
      Finished:     Tue, 29 May 2018 09:32:58 -0400
    Ready:          False
    Restart Count:  1
    Limits:
      cpu:     0
      memory:  0
    Requests:
      cpu:     0
      memory:  0
    Environment:
      ROOK_NODE_NAME:      (v1:spec.nodeName)
      ROOK_PRIVATE_IPV4:   (v1:status.podIP)
      ROOK_PUBLIC_IPV4:    (v1:status.podIP)
```

```
    Command:
      ceph-osd
    Args:
      --foreground
      --id
      1
      --conf
      /var/lib/rook/osd1/rook.config
      --osd-data
      /var/lib/rook/osd1
      --keyring
      /var/lib/rook/osd1/keyring
      --cluster
      rook
      --osd-uuid
      a38df6aa-da9c-440a-8c8c-e59aadb9d83b
      --public-addr
      $(ROOK_PUBLIC_IPV4)
      --cluster-addr
      $(ROOK_PRIVATE_IPV4)
      
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Tue, 29 May 2018 09:37:47 -0400
      Finished:     Tue, 29 May 2018 09:37:47 -0400
    Ready:          False
    Restart Count:  3
    Limits:
      cpu:     0
      memory:  0
    Requests:
      cpu:     0
      memory:  0
    Environment:
      ROOK_NODE_NAME:      (v1:spec.nodeName)
      ROOK_PRIVATE_IPV4:   (v1:status.podIP)
      ROOK_PUBLIC_IPV4:    (v1:status.podIP)
```